datafusion/common/src/error.rs

```diff
@@ -359,84 +380,49 @@ impl DataFusionError {
-                        if let Some(inner) = inner.downcast_ref::<Self>() {
-                            checkpoint = inner;
-                            continue;
-                        }
-
-                        if let Some(inner) = inner.downcast_ref::<ArrowError>() {
-                            other_e = Some(OtherErr::Arrow(inner));
-                            continue;
-                        }
-
-                        // some errors are wrapped into `Arc`s to share them with multiple receivers
-                        if let Some(inner) = inner.downcast_ref::<Arc<Self>>() {
-                            checkpoint = inner.as_ref();
-                            continue;
-                        }
-
-                        if let Some(inner) = inner.downcast_ref::<Arc<ArrowError>>() {
-                            other_e = Some(OtherErr::Arrow(inner.as_ref()));
-                            continue;
-                        }
-                    }
-                }
-
-                // dead end?
-                break;
-            }
-
-            // traverse context chain
-            if let Self::Context(_msg, inner) = checkpoint {
-                checkpoint = inner;
-                continue;
+        // Note: This is a non-recursive algorithm so we do not run
+        // out of stack space, even for long error chains.
+
+        let mut last_datafusion_error = self;
+        let mut root_error: &dyn Error = self;
+        while let Some(source) = find_source(root_error) {
+            // walk the next level
+            root_error = source;
+            // remember the lowest datafusion error so far
+            if let Some(e) = root_error.downcast_ref::<DataFusionError>() {
+                last_datafusion_error = e;
             }
-
-            // The Arrow error may itself contain a datafusion error again
-            // See https://github.com/apache/arrow-datafusion/issues/4172
-            if let Self::ArrowError(inner) = checkpoint {
-                other_e = Some(OtherErr::Arrow(inner));
-                continue;
-            }
-
-            // also try to introspect direct external errors
-            if let Self::External(inner) = checkpoint {
-                other_e = Some(OtherErr::Dyn(inner.as_ref()));
-                continue;
-            }
-
-            // no more traversal
-            break;
         }
-
         // return last checkpoint (which may be the original error)
-        checkpoint
+        last_datafusion_error
+    }
+}
+
+fn find_source<'a>(e: &'a (dyn Error + 'static)) -> Option<&'a (dyn Error + 'static)> {
+    // workaround until https://github.com/apache/arrow-rs/issues/3566 is released
+    if let Some(e) = e.downcast_ref::<ArrowError>() {
+        return if let ArrowError::ExternalError(e) = e {
+            Some(e.as_ref())
+        } else {
+            None
+        };
     }
+    // some errors are wrapped into `Arc`s to share them with multiple
```

Comment:
Is this necessary? The source implementation of Arc should already call this?

https://doc.rust-lang.org/src/alloc/sync.rs.html#2782

Reply:
It is necessary, but I don't know why. The tests fail without it. If you could help figure out how to avoid it it would be great.

Reply:
Fixed in https://github.com/apache/arrow-datafusion/pull/5056/commits/f3d2ae644f57cac7fd06f1d1c565ced765e42182